crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,709 @@
+            .await
+            .map_err(|e| PutSnapshotError::from(e, progress.clone()))?;
+        progress.push("/metadata.json".to_owned());
+
+        debug!(
+            etag = put_result.e_tag.unwrap_or_default(),
+            key = ?metadata_key,
+            "Successfully published snapshot metadata",
+        );
+
+        let latest_path = object_store::path::Path::from(format!(
+            "{prefix}{partition_id}/latest.json",
+            prefix = self.prefix,
+            partition_id = snapshot.partition_id,
+        ));
+
+        // By performing a CAS on the latest snapshot pointer, we can ensure strictly monotonic updates.
+        let maybe_stored = self
+            .get_latest_snapshot_metadata_for_update(snapshot, &latest_path)
+            .await
+            .map_err(|e| PutSnapshotError::from(e, progress.clone()))?;
+        if maybe_stored.as_ref().is_some_and(|(latest_stored, _)| {
+            latest_stored.min_applied_lsn >= snapshot.min_applied_lsn
+        }) {
+            let repository_latest_lsn = maybe_stored.expect("is some").0.min_applied_lsn;
+            info!(
+                ?repository_latest_lsn,
+                new_snapshot_lsn = ?snapshot.min_applied_lsn,
+                "The newly uploaded snapshot is no newer than the already-stored latest snapshot, will not update latest pointer"
+            );
+            return Ok(());
+        }
+
+        // Construct the new "latest snapshot" pointer
+        let latest = LatestSnapshot {
+            version: snapshot.version,
+            cluster_name: snapshot.cluster_name.clone(),
+            node_name: snapshot.node_name.clone(),
+            partition_id: snapshot.partition_id,
+            snapshot_id: snapshot.snapshot_id,
+            created_at: snapshot.created_at.clone(),
+            min_applied_lsn: snapshot.min_applied_lsn,
+            path: relative_snapshot_path,
+        };
+        let latest_json = PutPayload::from(
+            serde_json::to_string_pretty(&latest)
+                .map_err(|e| PutSnapshotError::from(e, progress.clone()))?,
+        );
+
+        // The object_store file provider supports create-if-not-exists but not update-version on put
+        let use_conditional_update = !matches!(self.destination.scheme(), "file");
+        let conditions = maybe_stored
+            .map(|(_, version)| PutOptions {
+                mode: match use_conditional_update {
+                    true => PutMode::Update(version),
+                    false => PutMode::Overwrite,
+                },
+                ..PutOptions::default()
+            })
+            .unwrap_or_else(|| PutOptions {
+                mode: PutMode::Create,
+                ..PutOptions::default()
+            });
+
+        let put_result = self
+            .object_store
+            .put_opts(&latest_path, latest_json, conditions)
+            .await
+            .map_err(|e| PutSnapshotError::from(e, progress.clone()))?;
```

Comment:
Sorry for being a bit pedantic here but I don't understand how `object_store` could retry on concurrent modifications w/o us checking whether the concurrent modification has a higher or lower lsn than what we wanted to write and then to update the `conditions` with the updated `version` of the concurrent modification. I think what `object_store` does is to retry on 409 which might arise with concurrent modifications but it shouldn't succeed in completing if there was a new version written.

Reply:
Not at all! I missed some subtlety here - the current behavior is to bail on a conditional modification error, so the `CreateSnapshot` request overall fails. But really, we could retry the application-level logic when we get a `Precondition` error back from `object_store` (which is an S3 412 error). This is an optimization I didn't think worthwhile during development but can also add, maybe as a follow-up task.

What I was referring to is that the S3 PUT operation (which is made with an etag constructed from the earlier `latest.json` we found, against which the application-level LSN logic works) is retried by the `object_store` `S3Client` when we get a 404 or 409 error from S3. (Apparently both are possible under different concurrent modification conditions.)

Reply:
I don't think it would be too hard to add retries here, but it's not rally needed at this point IMO. I have added a comment about this - let me know if you disagree, and I can do it as a fast follow-up. It won't be too much code to add a retryable `SnapshotError` variant and handle it in `put_snapshot`.